crypto/aes/asm/aes-s390x.pl

```diff
@@ -2219,6 +2254,69 @@ ()
 .size	AES_xts_decrypt,.-AES_xts_decrypt
 ___
 }
+
+################
+# void s390x_aes_gcm_blocks(unsigned char *out, GCM128_CONTEXT *ctx,
+#                           const unsigned char *in, size_t len,
+#                           const unsigned char *aad, size_t alen,
+#                           const AES_KEY *key, int enc)
+{
+my ($out,$ctx,$in,$len,$aad,$alen,$key,$enc) = map("%r$_",(2..9));
+$code.=<<___;
+.globl	s390x_aes_gcm_blocks
+.type	s390x_aes_gcm_blocks,\@function
+.align	16
+s390x_aes_gcm_blocks:
+.cfi_startproc
+	stm$g	$alen,$enc,7*$SIZE_T($sp)
```

Comment:
Aren't you obliged to preserve %r6 per ABI specification? Even if it's used as argument...

Reply:
%r6 belongs to the caller. spec says: "A called function shall save these registers' values before it changes them, restoring their values before it returns.", but in this case the register's value is not changed by the callee.

Reply:
But kma spec says that it is modified [in case when $alen is non-zero]: "As part of the operation, the address in general register R3 is *incremented* by the number of third-operand bytes processed, and the length in general register R3 + 1 is decremented by the same number;"

Reply:
z14 kma implements aes-gcm as a whole, including aad processing and tag generation. So naturally it is best used in a "one-pass" way and the rationale behind the design was to minimze calls to the cpacf coprocessor, because there is some cycles overhead which especially hurts small block processing performance.

libcrypto api enforces an init-update-final call sequence, making it impossible to process AAD and PT/CT in the same cpacf call. This makes kma aad processing useless, since its not faster than kimd.

So with the present implementation, the function is always called with aadlen=0 and therefore aad ptr %r6 is not changed. Thinking about it again, it would make sense to completely remove the functions aad arguments (or at least zero aadlen explicitely (I wrote that code w/o z14 HW access thinking that kma aad processing would probably be prefered to kimd).

Reply:
Well, even if subroutine in question isn't called with aadlen!=0 *now*, it's no excuse to assume that it won't ever. Or in other words subroutine that accepts aadlen as argument should assume that it can be non-zero and be implemented accordingly. As for removal aad arguments. That would solve the problem, but question is do we have to slam the door? In other words I'd suggest to keep aad arguments, just not assume that aadlen is always zero :-)

Reply:
If i keep the aad args (and save r6), it would also be consequent to remove remove # -ing out of the setup of param block's length field and make the function enable "full kma" (instead of implementing only the kma functionality that is used).

Reply:
And you can give all the power to caller by making it pass Flags and Modifier fields instead of enc... Or why not *whole* GR0 value?

Reply:
would you actually prefer the "give all power to the caller" approach? technically you only need an "inline assembly" kma instruction (see e.g. kernel source) and the caller could set up even the param block. i implemented it this way only to be consistent with existing asm i.e. seperate asm modules instead of inline asm, store function code (plus flags) in key field of aes ctx, set up param block in the asm ... the decrypt bit is special in this case, because the implementation also uses "plain" aes code path for partial block processing (see other comment) which obviously must always be an "encrypt" operation.

Reply:
>Or in other words subroutine that accepts aadlen as argument should assume that it can be non-zero and be implemented accordingly.

I agree on this, just forgot to adapt the relevant code parts when i removed kma aad processing.
However, in general i think its okay for the asm modules to make certain assumptions about the context from which they are called, e.g. checking the corresponding capabilities in the c glue.

Reply:
> in general i think its okay for the asm modules to make certain assumptions about the context from which they are called

Yes, of course. And on same general note it's always a compromise, i.e. what is done in assembly and what's done in C. And it's multi-variable problem with no ready answer. So it's only natural that it's solved apparently differently in different situations. And it's only natural that it can be discussed on per-case basis...

Reply:
> would you actually prefer the "give all power to the caller" approach?

It was rather a question that popped into mind than actual suggestion... Speaking of compromises, here it's about how complicated would assumption logic have to be, right? I mean if assembly is supposed to figure out flags from arguments, then caller would have to be aware of it to drive assembly to desired result. It might be trivial (I haven't thought of that deep enough yet), or it might be appropriate to implement the logic completely in C skipping assumptions altogether...

Reply:
Looking ahead at the waiting ecb/cfb(8)/ofb/ccm patches, which are right now analoguously implemented to gcm (that is e_aes.c wrapper plus short aes-s390x.pl function):
I would say replacing the "perlasm" implementations by inline asm (plus param block wrapper) is the easiest option, from a c programmers point of view (thats how we do it in the kernel: https://elixir.free-electrons.com/linux/v4.14-rc8/source/arch/s390/include/asm/cpacf.h , kma code not upstream yet). aes-ctr would have to be an excetiption, i.e. the only kma call outside of the inline asm.
The reason i implemented it this way was because the idea of keeping all platform specific asm in "standalone" modules is also appealing, from the view of folks cloning libryptos asm and the asm-implemented logic is not too hard to get, from an assembler programmers view.

Reply:
Well, usual counter-argument is that inline assembler is compiler-specific, and even can be version-specific. Or in other words assembly modules give you much more freedom to reuse the code. One can argue that s390x is "mono-cultural" enough to put this argument aside. Yet there is z/OS. Though I don't actually know if assembly code is reusable there... In other words I [for one] would still prefer to have it as assembly subroutine. At least GCM and CCM, as ones that are significant to real-life applications such as TLS, significant enough to not leave it to chance such as dependency on specific compiler.

Reply:
okay, id be happy to agree on keeping things the in asm modules (less rework for me :-)). yet i see a point in making the asm subroutines more "inline-asm-like" i.e. giving the caller all flexibility of the cpacf instruction.